rust/connlib/libs/tunnel/src/ip_packet.rs

```diff
@@ -0,0 +1,216 @@
+use std::net::IpAddr;
+
+use domain::base::message::Message;
+use pnet::packet::{
+    ip::{IpNextHeaderProtocol, IpNextHeaderProtocols},
+    ipv4::{checksum, Ipv4Packet, MutableIpv4Packet},
+    ipv6::{Ipv6Packet, MutableIpv6Packet},
+    udp::{ipv4_checksum, ipv6_checksum, MutableUdpPacket, UdpPacket},
+    MutablePacket, Packet, PacketSize,
+};
+
+const DNS_PORT: u16 = 53;
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum MutableIpPacket<'a> {
+    MutableIpv4Packet(MutableIpv4Packet<'a>),
+    MutableIpv6Packet(MutableIpv6Packet<'a>),
+}
+
+// no std::mem:;swap? no problem
+macro_rules! swap_src_dst {
+    ($p:expr) => {
+        let src = $p.get_source();
+        let dst = $p.get_destination();
+        $p.set_source(dst);
+        $p.set_destination(src);
+    };
+}
+
+impl<'a> MutableIpPacket<'a> {
+    pub(crate) fn new(data: &mut [u8]) -> Option<MutableIpPacket> {
+        match data[0] >> 4 {
+            4 => MutableIpv4Packet::new(data).map(MutableIpPacket::from),
+            6 => MutableIpv6Packet::new(data).map(MutableIpPacket::from),
```

Comment:
```suggestion
    pub(crate) fn new(data: &mut [u8]) -> Option<Self> {
        match data[0] >> 4 {
            4 => MutableIpv4Packet::new(data).map(Into::into),
            6 => MutableIpv6Packet::new(data).map(Into::into),
```

Reply:
I rather not use a `Self` return type here, that disables the elision and requires a `&'a` for the `data parameter.

Reply:
Ah, I hadn't considered that, fair enough!